### ISSUE TITLE
More robust/standard configuration

### ISSUE THREAD

**Issue by olafurpg** [olafurpg/scalafmt#316]
Following discussions on Gitter: https://gitter.im/olafurpg/scalafmt?at=5755e80492fc7c915f574b06

The current configuration is built on scopt and is quite hacky: https://olafurpg.github.io/scalafmt/#Configuration
It would be great to use a more standard syntax.

Requirements:
* Scala.js compatible
* supports comments
* not verbose
* allows multiple configurations on same line, like `a=1;b=2`. See #315

Considered options:
* [shocon](https://github.com/unicredit/shocon) seems like a perfect fit. @andreaTP, are there any docs/non-SNAPSHOT releases?
* typesafe config: no Scala.js support.
* http://oncue.github.io/knobs/: can't see Scala.js support.
* YAML: Don't think there's any Scala.js implementation.

**Comment by andreaTP** [comment]
shocon is right now used in akka-js as a "drop in" replacement of typesafe config.
It is developed by @evacchi, it works, but it is not complete.
If you want to give a try use
```
"eu.unicredit" %%% "shocon" % "0.0.2-SNAPSHOT"
```
from 
```
resolvers += "sonatype-snapshots" at "https://oss.sonatype.org/content/repositories/snapshots"
```
We will be happy to support!

**Comment by djspiewak** [comment]
As a counter-point…  Hocon is an incredibly annoying configuration format.  :-)  It's better than JSON, but I've just had universally massively negative experiences with it.  YAML is very lightweight, and its only negative aspects are a) hard to write a fully-compliant parser, and b) association with Rails.

Verizon uses YAML internally quite extensively, and it appears that we primarily use the following dependency:

```sbt
libraryDependencies += "org.yaml" %  "snakeyaml" % "1.16"
```

Not the prettiest API to work with, but it does work.

Why does scalafmt need to be Scala.js compatible?

**Comment by olafurpg** [comment]
> Why does scalafmt need to be Scala.js compatible?

I want make a browser extension that formats Scala code on Github with scalafmt. If #315 gets implemented then then scalafmt-core will depend on the parser for the configuration language.

Does YAML support multiple statements on the same line? Like `a=1;b=2`.

**Comment by djspiewak** [comment]
> Does YAML support multiple statements on the same line? Like a=1;b=2.

Yes.  It's not common style, but it is allowed.  The following are equivalent:

```yaml
foo:
  a: 1
  b: 2

# and this

foo: { a: 1, b: 2 }
```

**Comment by olafurpg** [comment]
My biggest concern is bringing in a massive dependency on a configuration library when scalafmt only needs the most basic features. There's no need for dates, environment variables, system properties, imports/config loading from other files, automatic reloading on resource change (although that might be nice for the IntelliJ plugin), etc.

Thing is scalafmt configuration files should stay tiny, max 5 lines IMO. If you need more than that I think we should reconsider the provided styles. For large organizations like Verizon I'm happy to make `--style verizon` a thing.

**Comment by djspiewak** [comment]
I guess it depends on how much much more advancement we see in configurability.  Eclipse and IntelliJ both have very elaborate configuration schemas for their Java formatters, and Scala is a more syntactically complex language.

**Comment by olafurpg** [comment]
True, scalafmt may accumulate quite a big selection of options. Still, most users shouldn't deviate too far from a provided base style. I imagine this is the case for ClangFormat, which probably has 80+ options and an excellent selection of base styles.

I'm still leaning more towards shocon because of Scala.js compat (that browser extension is gonna be so cool 😄). From what I can tell, shocon only depends on fastparse, which scalafmt already depends on transitively via scala.meta for XML literal parsing.

**Comment by djspiewak** [comment]
I definitely can't offer you a YAML parser that is even remotely Scala.js compatible.  Maybe one will happen in the future, but many things can happen in the future.

**Comment by olafurpg** [comment]
We probably won't see this issue implemented until https://github.com/scalameta/scalameta/issues/359 is fixed, which may take a while. There is still hope :)

**Comment by ngbinh** [comment]
👍 for YAML or even JS config.

**Comment by sjrd** [comment]
I would suggest YAML, and find a JavaScript implementation of YAML to use from Scala.js when the time comes.

**Comment by olafurpg** [comment]
Seems like YAML is popular, I'm happy to reconsider.

@sjrd makes an interesting suggestion, we still don't even know if/when scala.meta will support Scala.js. When the time comes, it seems like js-yaml could solve our problems: http://nodeca.github.io/js-yaml/ 

I also saw suggested on twitter that it's possible use snakeyaml via [jackson](https://github.com/FasterXML/jackson-dataformat-yaml) via [circe](https://github.com/travisbrown/circe). Not sure if that's the way to go but it's worth a look. Thanks for the input everybody.

**Comment by andreaTP** [comment]
using YAML through a javascript lib means have different code in js/jvm config reading unfortunately...

**Comment by djspiewak** [comment]
Most YAML libraries parse into some sort of map.  The config reading code doesn't need to be that distinct, just the bit that actually talks to the parser.

**Comment by jvican** [comment]
No one is using it as far as I am concerned. As a result, some parts of the user API are not very polished. I may implement inline tables soon and fix those things if you settle on TOML.

**Comment by djspiewak** [comment]
@olafurpg Infinite proliferation of configuration languages…

TOML looks quite a bit syntactically heavier than YAML, at least to my eyes.  YAML obviously has crazier things in its specification that yield a much more complex system, but the core is very lightweight.

I'm *really* tempted now to specify a strict subset of YAML and dash off a Scala parser for it…

YAML also supports "inline tables": 

```yaml
key1: { key2: 1, key3: foo }

# which is equivalent to

key1:
  key2: 1
  key3: foo
```

**Comment by jvican** [comment]
I think that your *strict subset of YAML* is TOML :wink:. The very fact that the spec is 84 pages long hints that YAML is not an easy, simple configuration language. At least, I'm very sure you wouldn't be able to make a simple parser implementation.
  
Also, note that TOML is used by [Cargo](http://doc.crates.io/), Rust's package manager, and other succesful projects. Its expressivity is not limited in any sense. Different strokes for different folks, I guess!

**Comment by djspiewak** [comment]
> I think that your strict subset of YAML is TOML 😉. The very fact that the spec is 84 pages long hints that YAML is not an easy, simple configuration language. At least, I'm very sure you wouldn't be able to make a simple parser implementation.

Most of the spec deals with vagaries of recursive references and document splits and such.  Another large swath of the spec deals with the redundant syntaxes allowed (e.g. `:` vs `=`, where the latter is basically never used in idiomatic YAML).  I'm relatively certain that a spec covering the subset of YAML that people actually *use* (i.e. not the subset generated by Ruby's object marshaling) would be quite short.

TOML isn't actually a YAML subset, so far as I can tell.  It's relatively close, but has a decent set of features which diverge (e.g. the `[key.subkey]` stuff).

**Comment by andreaTP** [comment]
since you are preprocessing why not use Scala itself instead? :-P

**Comment by jvican** [comment]
Yes, you're right @djspiewak, most of the spec deals with redundant syntaxes which are allowed and should not be, in my opinion. It's also true that TOML is not really a YAML subset, but as you said it's pretty close to it, and to me it feels natural to read. It's a simplified version, with a little bit more of sugar, and a spec that fits in a README file. I'm not going to discuss which one is better, and I don't feel very strong about this. Everyone has their own preferences :smile:.

**Comment by olafurpg** [comment]
It seems that YAML is the most popular configuration choice so I now have a working implementation for reading YAML files as configuration, see example: https://github.com/olafurpg/scalafmt/blob/43c2b98e08ca487fa57995af6a3cf772aee6a0d5/cli/src/test/scala/org/scalafmt/cli/YamlSpec.scala#L13

I use [circe-yaml](https://github.com/jeremyrsmith/circe-yaml) to convert snakeyaml nodes into circe JSON and then circe-generic to automatically decode the JSON into my own `ConfigurationOptions` case class. The solution is not perfect since it requires quite a few heavy dependencies. However, I get a lot of things for free. For example, the parse error messages from snakeyaml and conversion errors messages from circe have been very good in my small experiment. In the end I care most about developer productivity and a good user experience for users.

A neat side-effect of this solution is that it's possible to support JSON configuration as well. It should also be quite easy to add Scala.js support further down the line, since the problem boils down to converting yaml to json.

All that's left to do is update the integrations to support `scalafmt.yml` files.

What do y'all think?

**Comment by andreaTP** [comment]
snakeyaml is available only on JVM ...
b.t.w. shocon 0.1.1 is available in release :-P

**Comment by djspiewak** [comment]
@olafurpg Some thoughts:

- The structure is very… flat.  At the very least, I would expect some headings like `alignment:`, `binPack:` and similar.
- The `alignTokens` map is rather confusing.  Not only does it use an array element (but only one) which is unlike the other maps, but the values don't make any sense to me.  What possible values could appear in the `owner:` key?
- `indentOperator` and `rewriteTokens` are similarly confusing. `from: to; to: from`??
- Idiomatically, YAML files generally use snake-case (all lower-case with underscores separating words) rather than camel case.  `.travis.yml` is a pretty good exemplar of idiomatic YAML.

**Comment by johnynek** [comment]
I use snakeyaml with circe, which give nice YAML with pretty nice scala API:

https://github.com/johnynek/bazel-deps/blob/master/src/scala/com/github/johnynek/bazel_deps/CirceYaml.scala

**Comment by olafurpg** [comment]
Small update.

I tried shocon and I think it worked pretty great to start with. However, I quickly got into trouble when I wanted to read arrays of objects into my own case classes. I really don't like working with java.util.* and by the end I had implemented a mini-library for quite a bit of functionality that's already baked into circe.

I continued my work with circe-yaml and I think eventually I will go with this approach. @djspiewak I addressed some of your comments, see here. I think configurable snake case mapping is on the roadmap for circe, I couldn't find any references to it on the master branch.

I am currently blocked by travisbrown/circe#374 . My experience is that users enter typos all the time and I think scalafmt should fail on an invalid configuration file instead of silently falling back to the default settings.

**Comment by olafurpg** [comment]
@djspiewak Actually, I prefer `alignTokens` over `align` as the former is more descriptive. I could easily rename the case class field.

A neat thing with circe is that it allows me to override the decoder for custom types which enables me to support more flexible syntax for `alignTokens`, see [here](https://github.com/olafurpg/scalafmt/blob/313999c182561acf2df7b1f55447d62d5d1504aa/cli/src/main/scala/org/scalafmt/cli/AlignTokenDecoder.scala#L8-L15).

> I believe these lines should be indented.

I guess.... I still don't really get the syntax for arrays in YAML ¯\_(ツ)_/¯

**Comment by djspiewak** [comment]
@olafurpg Well, I mean…

```yaml
align:
  tokens: true
  byCallSite: true
  byDefnSite: false
```

> I guess.... I still don't really get the syntax for arrays in YAML ¯\_(ツ)_/¯

It's indentation-significant, but it uses `-` (when in conjunction with newline) to determine that it's an array rather than a dictionary.  So if you're all on one line, it's literally JSON syntax: `[1, 2, 3, 4]`.  If you're spread across lines, you use an extra indent level and a hyphen (`-`) to denote each element:

```yaml
stuff1:
  - 1
  - 2
  - 3

stuff2: [1, 2, 3]
```

The `stuff1` and `stuff2` dictionary entries will contain the same data.  Removing the indent level from the `stuff1` example still *works* because the line following the colon (`:`) is empty, but it's a bit wonky.  Always better to use the appropriate significant indentation.

**Comment by olafurpg** [comment]
@djspiewak I've thought about refactoring the ScalafmtStyle to support hierarchical configuration like that. It might make it easier to understand how some of the settings are related.

Another fancy feature that we could in theory implement

```yaml
# 1st alternative
continuationIndent:
  callSite: 2
  defnSite: 2
align:
  openParen:
    callSite: false
    defnSite: false

# 2nd alternative
continuationIndent: 2
align:
  openParen: false
```

This could be supported if create a class like this

```scala
case class Location[T](callSite: T, defnSite: T)
```

and then define a custom `Decoder[Location[T: Decoder]]`.

Thinking about this is fun and all. In the end it kind of goes against my original goal with scalafmt.

**Comment by djspiewak** [comment]
@olafurpg Your example snippet seems extremely cool, to be sure.  I guess, one step at a time.  :-)

**Comment by cvogt** [comment]
Or instead of bringing in some external data representation language, parsers, decorders, etc for configuration, just use Scala :).

I don't know scalafmt's api's, so here is an example using a fantasized one:

```
scala -cp scalafmt.jar FormatConfig.scala
```


```scala
// FormatConfig.scala
import scalafmt._
format(
  args,
  ScalafmtStyle(
    maxColumn = 80,
    binpack = Some(...),
    continuationIndent = None,
    align = Align(...),
    indentOperator = ...,
    rewrite = ...
  )
)
```

**Comment by olafurpg** [comment]
@cvogt this approach gets tricky for example with integrations like the IntelliJ plugin. Either everyone would need to manually install a zip with custom classfiles or scalafmt would need to add a dependency on scala-compiler to support loading the configuration from the plugin. Validating the file would also be a pain.

**Comment by cvogt** [comment]
@olafurpg Yep, there can be challenges with this approach, because people don't usually do it. I am not sure if we are ready for this step as an ecosystem yet. But it might also save from a lot of pain that usually exists. It might be worth checking if there are simple ways to make it work for the cases at hand. E.g. can't IntelliJ call scalafmt programmatically in the first place, i.e. not needing any serialization format?

Btw, you could even implement a FormatConfig.scala to read from a YAML file and ship it with scalafmt ;).

**Comment by paulp** [comment]
Never let it be said we refused to over-engineer our five line configuration files. The parser for the configuration language has more rules than the parser for the language we are asking the formatter to format.

**Comment by mdedetrich** [comment]
Why not just use HOCON, there are two parsers already available (the typesafe one and also a pure Scala one which is being developed, i.e. https://github.com/unicredit/shocon), and its pretty much the defacto configuration language for any Scala language. It also seems to support everything that you are asking for

Unless there is a really good use case for YAML (I can't see one), I would go for HOCON. Yaml is also notoriously hard to parse, and due to its really lax syntax it can be confusing to read config files in it (i.e. multiline arrays/enums)

TOML is a better alternative, but I would try and use HOCON.

Basically I agree with the others here, unless there is a **really** compelling reason why HOCON can't be used (i.e. it can't parse something that you are looking for) I would try and use it.

**Comment by djspiewak** [comment]
@mdedetrich This was discussed earlier on this issue thread.  I'm really not a fan of HOCON, in general.  The only reason it's the de facto "Scala configuration language" is because the Typesafe config library is so popular, not because of any particularly compelling benefit to it as a config language.  YAML is considerably more concise, if nothing else, and doesn't carry along with it spec-level ties to the JVM with `classpath()` includes and similar.

I'm not sure I buy the "confusing to read" argument w.r.t. YAML.  The only YAML files I've ever had a hard time reading were extremely poorly written, and in some cases were only valid due to parser bugs.  In fact, due to significant indentation, in my experience YAML files tend to be *better* structured than their average HOCON equivalents since the parser will actually force you to format things in a reasonable way.

There's no question that this whole debate (HOCON vs YAML vs TOML vs JSON vs pure-Scala vs whatever) is a ton of bikeshedding, and very definitely shaped by prior bias.

**Comment by paulp** [comment]
> The only reason it's the de facto "Scala configuration language" is because the Typesafe config library is so popular, not because of any particularly compelling benefit to it as a config language.

This is what used to be known as "begging the question" before that phrase was coopted by misusage. Why is the typesafe config library "so" popular? (I actually didn't realize it was, but if so, great.) I'd think because it addresses the lamest parts of JSON in a config-file context, and hasn't yet woken up one day trying to be a web browser or something. If it's "because typesafe" I will be surprised, because I doubt that rationale will hold up if we assess all typesafe projects for popularity.

I see hocon as the sensible choice.

**Comment by paulp** [comment]
> HOCON vs YAML vs TOML vs JSON vs pure-Scala

Configuration vs. Scala is not bikeshedding, it is right at the heart of the complexity management disaster we face every day. Among the first four, comparatively at least, it's bikeshedding.

**Comment by djspiewak** [comment]
> I see hocon as the sensible choice.

Now who's begging the question?  :-)

HOCON is not a wrong choice.  I'm not sure that means it's the "sensible" choice, seeing as that implies that all of the other choices are "insensible" in some fashion.  HOCON has the advantage of better parsers and more adoption in the Scala community, to be sure, at the cost of a more verbose syntax.

> If it's "because typesafe" I will be surprised

It's the same reason that Typesafe's scala-logging library is popular.  It's a core infrastructure framework, which means that people are going to gravitate toward the well-known, but not "core enough" that people gravitate toward the incredibly well-proven and trusted.  No one has ever been fired for choosing Typesafe Config.

> Configuration vs. Scala is not bikeshedding, it is right at the heart of the complexity management disaster we face every day.

Here we agree.

**Comment by mdedetrich** [comment]
@djspiewak 

> YAML is considerably more concise,

Conciseness is not always a good thing, it makes it much harder to pass and also introduces ambiguities. Also if you think the spec for HOCON is bad, you should see how complex the YAML spec is, especially with everything that they are adding to it

> if nothing else, and doesn't carry along with it spec-level ties to the JVM with classpath() includes and similar.
The `classpath()` isn't something that is specific to HOCON, and the Scala.js implementation (which has no concept of classpath) implements HOCON itself. Although on a spec level you could argue that the classpath is tied to HOCON, the syntax and parser itself isn't specifically tied to HOCON on a technical level

> There's no question that this whole debate (HOCON vs YAML vs TOML vs JSON vs pure-Scala vs whatever) is a ton of bikeshedding, and very definitely shaped by prior bias.

Its not really bikeshedding if you consider the economic cost of going for a configuration language that is already pretty standard in the Scala community. Also there is nothing stopping people from incrementally improving HOCON (i.e. removing the classpath dependency). There is however significant benefit if the community sticks (as much as possible) to a single configuration format. There of course may be exceptions (i.e. if you need to be really fast at parsing, or if your configuration language can't actually represent what you need it to) but that isn't the case here.

Not going to comment too much on the merits of the syntax, I think we will have to agree to disagree here (also do not think its that relevant because it comes down to taste)

**Comment by cvogt** [comment]
I know that suggesting Scala for configuration would likely be perceived as controversial. And I know I am not going to win anyone over right here and right now. But I would still love to place some of my thoughts on this topic in people's minds to at least give them a chance to agree or disagree with the specifics. It still seems to me like the right thing to do in the long run (maybe not for this project).

We currently mostly not use Scala but some other config language and often run into these problems:

- We need a replica of the toolchain for the config language: a parser, an interpreter, editor syntax highlighting, code formatting, etc.
  -   These may or may not work as well as our Scala ones regarding error messages, etc.
  -   These may break like all software, so more risk of stopping us solving the real problems.
  -   some parts of the tool chain often don't exist at all, especially static checking e.g. linting, type checking, syntax highlighting, etc. - so we accept "performance degredation"
  -   integration between Scala and the other language may be spotty, e.g. config files would not to be checked against the expected types in Scala
- Even configuration sometimes requires re-use. Config languages sometimes support some features here (e.g. HOCON variable substitution), but we sometimes run into situations where we have to copy and paste or use code generation with all the problems and complexities those bring.
- People are tempted to unsafely integrate on the file format with the same problems you have when you integrate software on a database - a hard time changing your mind about the structure (and the language here). Integrating through APIs seems much safer.
- we have to learn and remember a ton of config formats and their semantics

I think most of us immediately run into the lack of static checking and related crashes or manual labor but somehow accept this as necessary. This seems less of a problem for a tool like scalafmt, because it is run offline.

Other problems we don't run into so frequently, but when it we it is painful and seemingly unnecessary. If we would align and focus on one way to use the same language, tool chain, knowledge for even config and get the ecosystem ready for it, we'd probably be better off. But we are not going into this direction at the moment.

**Comment by paulp** [comment]
> Other problems we don't run into so frequently, but when it we it is painful and seemingly unnecessary.

I don't accept any of it as necessary - far, far from it - but the remedy is to stop treating configuration files like untyped blobs of characters, not to abandon configuration files entirely and add on to the codebase. Because that's what "configuration files in scala" means: code your configuration. Yes, the sledgehammer works, yes, some buildings require sledgehammers, but sledgehammers remain suboptimal for many, many tasks.

Principle of least power please.

**Comment by mdedetrich** [comment]
@cvogt I am not against configuration scalafmt using just code but it seems that @olafurpg wants to settle on a config format.

It might make more sense to have the core of scalafmt use scala code as config and then let the community to specify different config formats. This might make sense, because sbt has a different config style compared to typesafe HOCON (or even yaml if you want to use it).

Hell thats what the Java community seems to be doing with newer code (i.e. using builders as config)

**Comment by cvogt** [comment]
> stop treating configuration files like untyped blobs of characters
👍 

> code your configuration

Exactly. To the point of config file nirvana - they disappear ;). And yes this has practical problems right now, but not sure if any conceptual ones.

> Principle of least power please.

I understand this principle as using the least power when solving a problem yourself. Not as restricting the power of others. (Helping them choose is good, but locking them in not so much). Not sure yet if this is a comprehensive enough view.

**Comment by paulp** [comment]
@cvogt When you express your configuration in scala, only a tool which can understand scala can understand it - and only as well as scala can be understood (which of course statically is not very well.) We restrict the language of expression because of the commensurate gains in how, where, and with what level of difficulty we can be understood.

I do not need a general purpose programing language to configure my code formatter. Neither do you.

**Comment by cvogt** [comment]
>  When you express your configuration in scala, only a tool which can understand scala can understand it

That's only true if tools try to read that file, i.e. integrate on the file format. If the tool calls an api that hands it out in some format the tool understands there is no problem.

> I do not need a general purpose programing language to configure my code formatter. Neither do you.

If you see it in isolation - you are completely right. If you see everyone using a different config format with very limited guarantees - maybe I do.

**Comment by paulp** [comment]
@cvogt aren't you restricting the power of people who require macros to tell the code formatter two spaces or four? How are they supposed to get by with mere Scala?

**Comment by MasseGuillaume** [comment]
What about no configuration at all. For instance, with gofmt you can only configure tab size.

**Comment by mdedetrich** [comment]
@MasseGuillaume
> What about no configuration at all. For instance, with gofmt you can only configure tab size.

Easier said then done. Go is intentionally designed to be much less expressive (both in terms of syntax and power) compared to Scala, so its much easier to settle on a style of syntax. There for example is no operator overloading or defining of infix/postfix methods in Go, which makes formatting much easier

**Comment by olafurpg** [comment]
> [Scala as config ...]

It's a fascinating topic, no doubt. I'm personally in favor of the opposite trend. I would love to use an SBT/CBT plugin that allows me to setup simple but non-trivial builds in HOCON/YAML.

Scala as config language is absolute overkill for scalafmt so it's a no go.  Also, I have no idea how it would play with integrations like the intellij plugin that is distributed via plugins.jetbrains.com and upgraded automatically by the IDE.

> [No configuration ... gofmt ] 

I would love it if everyone could agree on a single style. Unfortunately, that would probably mean many scalafmt users would go back to manual formatting, which is a step backwards.

Saying that gofmt has no configuration option is only half of the story, IMO. Gofmt preserves line breaks leaving it up to developers and code reviewers to decide on the configuration (maxColumn/bin-packing) for every single line of code. Scalafmt (like clang-format/dartfmt/rustfmt/rfmt/...) is more strict on how you break lines and for that reason I think it's OK to let the user decide on a configuration once and then focus on getting work done.

I hope we will eventually consolidate on a small set of --style default/intellij/scala.js. At the moment we are still experimenting with what coding styles we want. For the record, every scalafmt config option outside of --style default has been requested/contributed by users, not myself.

I have a WIP to support HOCON since that seems to be what the Scala community is moving towards, for better or worse. I greatly appreciate everyone's feedback. I did my best to take your suggestions into account.

 From now on, please let's try to keep this thread focused on closing this issue.

**Comment by alexandru** [comment]
I personally hate YAML as a configuration format because it is white-space sensitive. I can understand the ability to skip `;` or `,` at the end of lines, but having your data-structures defined by your indentation level and sensitive to usage spaces versus tabs is horrible. Had many problems over the years because of that. I don't understand why people like it.

Personally I like TypeLevel Config.

**Comment by cvogt** [comment]
> setup simple but non-trivial builds in HOCON/YAML.

I am skeptical of scaling issues (i.e. requiring a language switch in more interesting cases), but would love to see an example of what that would look like.

> Scala as config language is absolute overkill for scalafmt so it's a no go.

Will try to whip up some example projects trying to make my case, unrelated to scalafmt. This topic has been on my mind for a while.

Going with HOCON sounds like a non-controversial, familiar, reasonably well understood choice. 👍 I'd appreciate loose coupling so it is possible to bypass the hocon loader and call scalafmt as a library providing config as an argument.

**Comment by mdedetrich** [comment]
@olafurpg 

> From now on, please let's try to keep this thread focused on closing this issue.

More on topic (and since it seems you have decided on HOCON), do you need the config to work for Scala.js as well or does it need to be JVM only?

If is JVM only, then https://github.com/typesafehub/config + https://github.com/iheartradio/ficus is what I would use. There is also https://oncue.github.io/knobs/ if you want something that is more purely functional.

Else you, shocon probably needs a bit of work to get it up to date

**Comment by SethTisue** [comment]
> Personally I like TypeLevel Config

@alexandru is there a folk we don't know about :-)

**Comment by dwijnand** [comment]
It partially unifies config files. ;-)

**Comment by andreaTP** [comment]
@mdedetrich shocon is already able to properly parse Akka standard config files, that are for sure magnitude orders much complex than what is needed here. Right now corner cases and config merges are not perfect but, please contribute! :-)

**Comment by olafurpg** [comment]
Small update.

I have a working [`@ConfigReader`](https://github.com/olafurpg/scalafmt/pull/459/files#diff-0c0a279d8c228c28f89fe35deb27252bR14) new style [inline/meta](https://github.com/scala/scala.github.com/pull/567) macro annotation that generates a `Reader[Map[String, Any]]`. The HOCON adapter is tiny: https://github.com/olafurpg/scalafmt/pull/459/files#diff-35211b88e3801c29d346c3dc04d7c46eR7 I think similar YAML/JSON/TOML adapters could easily be added, for that matter, but scalafmt will only support HOCON.

This gives me a configurable + 0 boilerplate + Scala.js compatible + Dotty compatible (!) config solution. I'm gonna start with com.typesafe.hocon but it it's comforting to know I can switch to shocon once meta compiles to Scala.js.

Sneak peak at an example scalafmt HOCON config file: https://github.com/olafurpg/scalafmt/pull/459/files#diff-035d8a3f2c8fa32d4e6333dfb183df46R38

**Comment by olafurpg** [comment]
I've released 0.4.0-SNAPSHOT that uses the new HOCON style configuration. Please take a look at it tell me if you were able to succesfully migrate the old configuration using the instructions here: https://github.com/olafurpg/scalafmt/releases/tag/0.4
